### PR TITLE
Adds a separate directory for adding external tilesets and exporting tilesets #453 

### PR DESCRIPTION
**Pull request by HenryJia**

### CODE REVIEW

File: src/tiled/tilesetdock.cpp [hunk 1]
```diff
@@ -758,20 +758,20 @@ void TilesetDock::exportTileset()
     if (!tileset)
         return;
 
-    const QLatin1String extension(".tsx");
-    QString suggestedFileName = QFileInfo(mMapDocument->fileName()).path();
-    suggestedFileName += QLatin1Char('/');
-    suggestedFileName += tileset->name();
-    if (!suggestedFileName.endsWith(extension))
-        suggestedFileName.append(extension);
```
Comment (bjorn) [inline src/tiled/tilesetdock.cpp]: Please don't just rip out the suggested file name logic. You'll need to combine the stored path with the suggested file name.

File: src/tiled/mainwindow.cpp [hunk 2]
```diff
@@ -1187,14 +1187,19 @@ void MainWindow::addExternalTileset()
     if (!mMapDocument)
         return;
 
-    const QString start = fileDialogStartLocation();
+    Preferences *prefs = Preferences::instance();
+
+    const QString start = prefs->lastPath(Preferences::ExternalTileset);
```
Comment (bjorn) [inline src/tiled/mainwindow.cpp]: Please still use `fileDialogStartLocation()` if the last path is empty (the first time people use it).

**Comment by bjorn** [comment]
Apart from my inline comment, please also remove the merge commit from the pull request.

**Comment by HenryJia** [comment]
All done.

**Comment by bjorn** [comment]
Why did you close this, @HenryJia?

**Comment by HenryJia** [comment]
Oops I think I must've accidentally deleted it whilst removing some branches that I created that weren't any use. I'll re open it

**Comment by HenryJia** [comment]
I've reopened it and restored the branch

File: src/tiled/mainwindow.cpp [hunk 3]
```diff
@@ -1187,14 +1187,23 @@ void MainWindow::addExternalTileset()
     if (!mMapDocument)
         return;
 
-    const QString start = fileDialogStartLocation();
+    Preferences *prefs = Preferences::instance();
+
+    QString start = prefs->lastPath(Preferences::ExternalTileset);
+
+    if(start.isEmpty())
```
Comment (bjorn) [inline src/tiled/mainwindow.cpp]: Please put a space between the `if` and `(`.

File: src/tiled/preferences.cpp [hunk 4]
```diff
@@ -340,6 +340,9 @@ static QString lastPathKey(Preferences::FileType fileType)
     case Preferences::ExportedFile:
         key.append(QLatin1String("ExportedFile"));
         break;
+    case Preferences::ExternalTileset:
+        key.append(QLatin1String("External Tileset"));
```
Comment (bjorn) [inline src/tiled/preferences.cpp]: Please remove the space from the settings key.

File: src/tiled/tilesetdock.cpp [hunk 5]
```diff
@@ -758,20 +758,31 @@ void TilesetDock::exportTileset()
...
+    QString suggestedFileName = prefs->lastPath(Preferences::ExternalTileset);
+
+    for(int i = (suggestedFileName.length() - 1); suggestedFileName.at(i) != QLatin1Char('/')
+        && suggestedFileName != empty; i--)
+        suggestedFileName.remove(i, 1);
+
+    if(suggestedFileName == empty)
+        suggestedFileName += QLatin1Char('/');
     suggestedFileName += tileset->name();
-    if (!suggestedFileName.endsWith(extension))
```
Comment (bjorn) [inline src/tiled/tilesetdock.cpp]: Why did you remove this condition? I think it should stay to avoid adding the extension twice when it is part of the tileset name.

File: src/tiled/tilesetdock.cpp [hunk 6]
```diff
@@ -758,20 +758,31 @@ void TilesetDock::exportTileset()
...
-    suggestedFileName += QLatin1Char('/');
+    QLatin1String empty(".");
+    QString suggestedFileName = prefs->lastPath(Preferences::ExternalTileset);
+
+    for(int i = (suggestedFileName.length() - 1); suggestedFileName.at(i) != QLatin1Char('/')
+        && suggestedFileName != empty; i--)
+        suggestedFileName.remove(i, 1);
+
+    if(suggestedFileName == empty)
+        suggestedFileName += QLatin1Char('/');
```
Comment (bjorn) [inline src/tiled/tilesetdock.cpp]: What does this do? If the suggestedFileName equals ".", it will add a slash? And what happened to the code that would suggest using the path of the map file? I think that should stay in case no path was saved yet for `ExternalTileset`.

File: src/tiled/tilesetdock.cpp [hunk 7]
```diff
@@ -758,20 +758,31 @@ void TilesetDock::exportTileset()
...
+
     const QLatin1String extension(".tsx");
-    QString suggestedFileName = QFileInfo(mMapDocument->fileName()).path();
-    suggestedFileName += QLatin1Char('/');
+    QLatin1String empty(".");
+    QString suggestedFileName = prefs->lastPath(Preferences::ExternalTileset);
+
+    for(int i = (suggestedFileName.length() - 1); suggestedFileName.at(i) != QLatin1Char('/')
+        && suggestedFileName != empty; i--)
+        suggestedFileName.remove(i, 1);
```
Comment (bjorn) [inline src/tiled/tilesetdock.cpp]: Please just keep the original approach that was using `QFileInfo::path`, because this is a little strange.

Comment (HenryJia) [inline src/tiled/tilesetdock.cpp]: The for loop is required because when you suggest a path using the lastPath it will also include the file name of the last path inside the string. The foor loop loops from the end until it reaches a / to remove that and then add the Tileset's name to it.

What do you mean by that? How do you want me to change it? I don't really see an easier way to do it.

Comment (bjorn) [inline src/tiled/tilesetdock.cpp]: I would prefer if you changed it back to the way it was done before:
```
suggestedFileName = QFileInfo(suggestedFileName).path();
```

File: src/tiled/tilesetdock.cpp [hunk 8]
```diff
@@ -758,20 +758,33 @@ void TilesetDock::exportTileset()
...
+
+    if(suggestedFileName == empty)
+        QString suggestedFileName = QFileInfo(mMapDocument->fileName()).path();
+        suggestedFileName += QLatin1Char('/');
```
Comment (bjorn) [inline src/tiled/tilesetdock.cpp]: The `if` condition here seems broken. Why do you expect the `suggestedFileName` to be "."? I think you want to put this check above the "extracting of the path" part, and just use `suggestedFileName.isEmpty()`.

**Comment by bjorn** [comment]
I've fixed up the commit a little and pushed it as 38932930c4c7bfb6ae0662e0b5c553632b5eedbd.

I'm sorry about some of my earlier comments, since they have not been entirely correct. In hindsight, it was not necessary to fall back to the current map file path since `Preferences::lastPath` was already doing this.

Thanks for your contribution!

**Comment by HenryJia** [comment]
Your welcome!
Thanks for letting me work on this too, I've learnt quite a lot from working on this project.
